recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,269 @@
+      "test.azurecr.io":{
+        "secret": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/secretStores/acrsecrets"
+      },
+    }
+  }
+  "env": {
+    ...
+  }
+},
+```
+
+#### Different container registry providers provide different authenticate methods to login.
+
+***Azure***:
+
+Azure provides multiple ways to authenticate ACR using ORAS, we can use different types of credentials depending on the setup and preferences. Here are the main types of credentials to use:
+
+#### Admin User Credentials:
+Users can enable the admin user and obtain these credentials from the Azure portal:
+```
+Username : The ACR admin username.
+Password : The ACR admin password.
+```
+
+#### Service Principal:
+```
+Username: The service principal's client ID.
+Password: The service principal's client secret.
+```
+
+#### Azure CLI Token:
+```
+Username: Use '00000000-0000-0000-0000-000000000000' (a placeholder indicating token-based authentication).
+Password: Use the token obtained from az acr login.
+```
+
+You can obtain the token using the Azure CLI:
+```
+az acr login --name <registry-name> --expose-token
+```
+***Github***:
+Github uses personal access token as password to auth into ghcr.
+```
+Username: Github username.
+Password: Personal access token.
+```
+
+***AWS***:
+```
+Username: AWS
+Password: output of `aws ecr get-login-password --region region`
+```
+
+### API design (if applicable)
+***Model changes***
+
+environment.tsp
+```diff
+@doc("Configuration for Recipes. Defines how each type of Recipe should be configured and run.")
+model RecipeConfigProperties {
+  @doc("Configuration for Terraform Recipes. Controls how Terraform plans and applies templates as part of Recipe deployment.")
+  terraform?: TerraformConfigProperties;
+
++  @doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++  bicep?: BicepConfigProperties;
+
+  @doc("Environment variables injected during Terraform Recipe execution for the recipes in the environment.")
+  env?: EnvironmentVariables;
+}
+
++@doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++model BicepConfigProperties {
++
++  @doc("Authentication information used to access private bicep registries, which is a map of registry hostname to secret config that contains credential information.")
++  authentication?: Record<SecretConfig>;
+}
+
+@doc("Personal Access Token (PAT) configuration used to authenticate to Git platforms.")
+model SecretConfig {
+  @doc("Secret represents the resource ID of the secret store that has credential information.")
+  secret?: string;
+}
+
+```
+***Bicep Example***
+
+environment.bicep
+```
+"recipeConfig": {
+  "terraform": {
+    ...
+  },
+  "bicep": {
+    "authentication":{
+      "test.azurecr.io":{
+        "secret": "/planes/radius/local/resourcegroups/default/providers/Applications.Core/secretStores/acr-secret"
```

Comment:
Will anything in this design need to change when we add AZWI or IRSA support? How will we tell the difference between the password case and AZWI casE?